File: grpc/golang/node.go
package freeplane_grpc

import (
	"context"
	"fmt"
)

// Node represents a node in a Freeplane mind map.
type Node struct {
	client  *FreeplaneClient
	nodeID  string
	mindMap *MindMap
	ctx     context.Context
}

// NewNode creates a new Node instance.
func NewNode(client *FreeplaneClient, nodeID string, mindMap *MindMap) *Node {
	return &Node{
		client:  client,
		nodeID:  nodeID,
		mindMap: mindMap,
		ctx:     context.Background(),
	}
}

// WithContext returns a copy of the Node with the provided context.
func (n *Node) WithContext(ctx context.Context) *Node {
	return &Node{
		client:  n.client,
		nodeID:  n.nodeID,
		mindMap: n.mindMap,
		ctx:     ctx,
	}
}

// ==================== Text Operations ====================

// GetText returns the text of the node.
func (n *Node) GetText() (string, error) {
	resp, err := n.client.GetNodeText(n.ctx, n.nodeID)
	if err != nil {
		return "", err
	}
	return resp.Text, nil
}

// SetText sets the text of the node.
func (n *Node) SetText(text string) error {
	_, err := n.client.SetNodeText(n.ctx, n.nodeID, text)
	return err
}

// ==================== Hierarchy Operations ====================

// AddChild creates a child node with the given text and optional style.
func (n *Node) AddChild(text, style string) (*Node, error) {
	resp, err := n.client.CreateChild(n.ctx, text, n.nodeID)
	if err != nil {
		return nil, err
	}
	return NewNode(n.client, resp.NodeId, n.mindMap), nil
}

// Children returns the list of child nodes.
func (n *Node) Children() ([]*Node, error) {
	resp, err := n.client.ListChildNodes(n.ctx, n.nodeID)
	if err != nil {
		return nil, err
	}
	var children []*Node
	for _, child := range resp.Children {
		children = append(children, NewNode(n.client, child.NodeId, n.mindMap))
	}
	return children, nil
}

// Parent returns the parent node.
func (n *Node) Parent() (*Node, error) {
	resp, err := n.client.GetParentNode(n.ctx, n.nodeID)
	if err != nil {
		return nil, err
	}
	if resp.NodeId == "" {
		return nil, NewNodeNotFoundError("node has no parent")
	}
	return NewNode(n.client, resp.NodeId, n.mindMap), nil
}

// Delete deletes the node.
func (n *Node) Delete() error {
	_, err := n.client.DeleteChild(n.ctx, n.nodeID)
	return err
}

// Move moves the node to a new parent.
func (n *Node) Move(newParentID string) error {
	_, err := n.client.MoveNode(n.ctx, n.nodeID, newParentID)
	return err
}

// ==================== Styling Operations ====================

// GetStyle returns the style name of the node.
func (n *Node) GetStyle() (string, error) {
	// Style is retrieved via Groovy script since there's no dedicated RPC
	resp, err := n.client.Groovy(n.ctx, fmt.Sprintf("return node.getStyle().toString()"))
	if err != nil {
		return "", err
	}
	return resp.Result, nil
}

// SetStyle sets the style name of the node.
func (n *Node) SetStyle(style string) error {
	// Escape single quotes to prevent Groovy injection
	escaped := style
	for i := 0; i < len(escaped); i++ {
		if escaped[i] == '\'' {
			escaped = escaped[:i] + "\\'" + escaped[i+1:]
			i++
		}
	}
	_, err := n.client.Groovy(n.ctx, fmt.Sprintf("node.getStyle().setStyle('%s')", escaped))
	return err
}

// GetColor returns the text color as RGBA.
func (n *Node) GetColor() (red, green, blue, alpha int32, err error) {
	// Color is retrieved via Groovy script
	resp, err := n.client.Groovy(n.ctx, "return node.style.color.toString()")
	if err != nil {
		return 0, 0, 0, 0, err
	}
	// Parse the color string (format: "r,g,b,a" or similar)
	// TODO: Implement actual color parsing from Groovy result
	if resp.Result == "" {
		return 0, 0, 0, 0, NewOperationError("empty color result from Groovy script")
	}
	_ = resp
	return 0, 0, 0, 0, nil
}

// SetColor sets the text color.
func (n *Node) SetColor(red, green, blue, alpha int32) error {
	_, err := n.client.NodeColorSet(n.ctx, n.nodeID, red, green, blue, alpha)
	return err
}

// SetBackgroundColor sets the background color.
func (n *Node) SetBackgroundColor(red, green, blue, alpha int32) error {
	_, err := n.client.NodeBackgroundColorSet(n.ctx, n.nodeID, red, green, blue, alpha)
	return err
}

// ==================== Notes Operations ====================

// GetNotes returns the note content of the node.
func (n *Node) GetNotes() (string, bool, error) {
	resp, err := n.client.GetNodeNote(n.ctx, n.nodeID)
	if err != nil {
		return "", false, err
	}
	return resp.Note, resp.HasNote, nil
}

// SetNotes sets the note content of the node.
func (n *Node) SetNotes(notes string) error {
	_, err := n.client.NodeNoteSet(n.ctx, n.nodeID, notes)
	return err
}

// ==================== Attributes Operations ====================

// GetAttributes returns the node attributes.
func (n *Node) GetAttributes() (map[string]string, error) {
	// Attributes are retrieved via Groovy script
	resp, err := n.client.Groovy(n.ctx, "def attrs = [:]; node.attributes.each { k, v -> attrs[k] = v.toString() }; return attrs")
	if err != nil {
		return nil, err
	}
	// Parse the result - for now return empty map
	if resp.Result == "" {
		return nil, NewOperationError("empty attributes result from Groovy script")
	}
	_ = resp
	return make(map[string]string), nil
}

// SetAttribute sets a single attribute.
func (n *Node) SetAttribute(name, value string) error {
	_, err := n.client.NodeAttributeAdd(n.ctx, n.nodeID, name, value)
	return err
}

// SetAttributes sets multiple attributes.
func (n *Node) SetAttributes(attrs map[string]string) error {
	for name, value := range attrs {
		if err := n.SetAttribute(name, value); err != nil {
			return err
		}
	}
	return nil
}

// ==================== Links Operations ====================

// GetLinks returns the node links.
func (n *Node) GetLinks() (string, bool, error) {
	resp, err := n.client.GetNodeLink(n.ctx, n.nodeID)
	if err != nil {
		return "", false, err
	}
	return resp.Link, resp.HasLink, nil
}

// SetLinks sets the node link.
func (n *Node) SetLinks(link string) error {
	_, err := n.client.NodeLinkSet(n.ctx, n.nodeID, link)
	return err
}

// ==================== Tags Operations ====================

// SetTags sets tags on the node (replaces existing tags).
func (n *Node) SetTags(tags []string) error {
	_, err := n.client.NodeTagSet(n.ctx, n.nodeID, tags)
	return err
}

// AddTags adds tags to the node (appends to existing tags).
func (n *Node) AddTags(tags []string) error {
	_, err := n.client.NodeTagAdd(n.ctx, n.nodeID, tags)
	return err
}

// ==================== Icons Operations ====================

// GetIcons returns the node icons.
func (n *Node) GetIcons() ([]string, error) {
	// Icons are retrieved via Groovy script
	resp, err := n.client.Groovy(n.ctx, "return node.iconNames.toArray()")
	if err != nil {
		return nil, err
	}
	if resp.Result == "" {
		return nil, NewOperationError("empty icons result from Groovy script")
	}
	_ = resp
	return nil, nil
}

// AddIcon adds an icon to the node.
func (n *Node) AddIcon(name string) error {
	_, err := n.client.NodeAddIcon(n.ctx, n.nodeID, name)
	return err
}

// ==================== State Operations ====================

// GetFolded returns whether the node is folded.
func (n *Node) GetFolded() (bool, error) {
	resp, err := n.client.Groovy(n.ctx, "return node.isExpanded()")
	if err != nil {
		return false, err
	}
	if resp.Result == "" {
		return false, NewOperationError("empty folded state result from Groovy script")
	}
	_ = resp
	return false, nil
}

// SetFolded sets the folded state of the node.
func (n *Node) SetFolded(folded bool) error {
	var cmd string
	if folded {
		cmd = "node.performCollapse()"
	} else {
		cmd = "node.performExpand()"
	}
	_, err := n.client.Groovy(n.ctx, cmd)
	return err
}

// ==================== Action Operations ====================

// Select selects the node.
func (n *Node) Select() error {
	return n.client.FocusNodeByID(n.ctx, n.nodeID)
}

// Centers the node in the view.
func (n *Node) Center() error {
	_, err := n.client.Groovy(n.ctx, "node.centerInView()")
	return err
}

// Refresh refreshes the node display.
func (n *Node) Refresh() error {
	_, err := n.client.Groovy(n.ctx, "node.refresh()")
	return err
}

// ==================== Properties ====================

// Client returns the FreeplaneClient.
func (n *Node) Client() *FreeplaneClient {
	return n.client
}

// NodeID returns the node ID.
func (n *Node) NodeID() string {
	return n.nodeID
}

// MindMap returns the parent MindMap.
func (n *Node) MindMap() *MindMap {
	return n.mindMap
}

// String returns a string representation of the node.
func (n *Node) String() string {
	text, _ := n.GetText()
	return fmt.Sprintf("Node{id=%s, text=%q}", n.nodeID, text)
}

File: grpc/cpp/src/node.cpp
#include "node.h"
#include "client.h"
#include "mindmap.h"

#include <sstream>

namespace freeplane {
namespace grpc {

Node::Node(FreeplaneClient* client, const std::string& nodeId, MindMap* mindmap)
    : client_(client), node_id_(nodeId), mindmap_(mindmap) {
}

// -- text ----------------------------------------------------------------

std::string Node::getText() const {
    auto resp = client_->getNodeText(node_id_);
    if (!resp->success()) {
        throw FreeplaneOperationError(
            !resp->error_message().empty() ? resp->error_message() : "Failed to get node text");
    }
    return resp->text();
}

void Node::setText(const std::string& text) {
    auto resp = client_->setNodeText(node_id_, text);
    if (!resp->success()) {
        throw FreeplaneOperationError(
            !resp->error_message().empty() ? resp->error_message() : "Failed to set node text");
    }
}

// -- hierarchy -----------------------------------------------------------

std::shared_ptr<Node> Node::addChild(const std::string& text, const std::string& style) {
    (void)style;  // Style is not directly supported via gRPC, uses default
    
    auto resp = client_->createChild(text, node_id_);
    auto node = std::make_shared<Node>(client_, resp->node_id(), mindmap_);
    
    // Update the mindmap's node_id to point to the new child
    if (mindmap_) {
        // MindMap doesn't expose mutable node_id, but we can note this
    }
    
    return node;
}

std::vector<std::shared_ptr<Node>> Node::children() {
    auto resp = client_->listChildNodes(node_id_);
    if (!resp->success()) {
        throw FreeplaneOperationError(
            !resp->error_message().empty() ? resp->error_message() : "Failed to list children");
    }
    
    std::vector<std::shared_ptr<Node>> result;
    for (const auto& child : resp->children()) {
        result.push_back(std::make_shared<Node>(client_, child.node_id(), mindmap_));
    }
    return result;
}

std::shared_ptr<Node> Node::parent() {
    auto resp = client_->getParentNode(node_id_);
    if (!resp->success() || resp->parent_node_id().empty()) {
        throw NodeNotFoundError("Node " + node_id_ + " has no parent (is root)");
    }
    return std::make_shared<Node>(client_, resp->parent_node_id(), mindmap_);
}

bool Node::deleteNode() {
    auto resp = client_->deleteChild(node_id_);
    return resp->success();
}

bool Node::move(const std::string& newParentId) {
    auto resp = client_->moveNode(node_id_, newParentId);
    if (!resp->success()) {
        throw FreeplaneOperationError(
            !resp->error_message().empty() ? resp->error_message() : "Failed to move node");
    }
    return resp->success();
}

// -- styling -------------------------------------------------------------

bool Node::setStyle(const std::string& style) {
    // Use Groovy to set style
    std::ostringstream groovy;
    groovy << "def node = model.getNode('" << node_id_ << "');"
           << "node.style = model.getStyleLib().getStyle('" << style << "');";
    
    auto result = client_->groovy(groovy.str());
    return result.find("Error") == std::string::npos;
}

bool Node::setColor(int32_t red, int32_t green, int32_t blue, int32_t alpha) {
    auto resp = client_->nodeColorSet(node_id_, red, green, blue, alpha);
    return resp->success();
}

bool Node::setBackgroundColor(int32_t red, int32_t green, int32_t blue, int32_t alpha) {
    auto resp = client_->nodeBackgroundColorSet(node_id_, red, green, blue, alpha);
    return resp->success();
}

// -- notes ---------------------------------------------------------------

std::pair<bool, std::string> Node::getNote() {
    auto resp = client_->getNodeNote(node_id_);
    if (!resp->success()) {
        throw FreeplaneOperationError(
            !resp->error_message().empty() ? resp->error_message() : "Failed to get node note");
    }
    return {resp->has_note(), resp->note()};
}

void Node::setNote(const std::string& note) {
    auto resp = client_->nodeNoteSet(node_id_, note);
    if (!resp->success()) {
        throw FreeplaneOperationError("Failed to set node note");
    }
}

// -- attributes ----------------------------------------------------------

bool Node::setAttribute(const std::string& name, const std::string& value) {
    auto resp = client_->nodeAttributeAdd(node_id_, name, value);
    return resp->success();
}

bool Node::setAttributes(const std::vector<std::pair<std::string, std::string>>& attrs) {
    for (const auto& attr : attrs) {
        if (!setAttribute(attr.first, attr.second)) {
            return false;
        }
    }
    return true;
}

// -- links ---------------------------------------------------------------

std::vector<std::string> Node::getLinks() {
    auto resp = client_->getNodeLink(node_id_);
    if (!resp->success()) {
        throw FreeplaneOperationError(
            !resp->error_message().empty() ? resp->error_message() : "Failed to get node link");
    }
    
    std::vector<std::string> result;
    if (resp->has_link()) {
        result.push_back(resp->link());
    }
    return result;
}

void Node::setLinks(const std::vector<std::string>& links) {
    for (const auto& link : links) {
        client_->nodeLinkSet(node_id_, link);
    }
}

// -- tags ----------------------------------------------------------------

bool Node::setTags(const std::vector<std::string>& tags) {
    auto resp = client_->nodeTagSet(node_id_, tags);
    return resp->success();
}

bool Node::addTags(const std::vector<std::string>& tags) {
    auto resp = client_->nodeTagAdd(node_id_, tags);
    return resp->success();
}

// -- icons ---------------------------------------------------------------

bool Node::addIcon(const std::string& iconName) {
    auto resp = client_->nodeAddIcon(node_id_, iconName);
    return resp->success();
}

// -- state ---------------------------------------------------------------

bool Node::getFolded() {
    // Use Groovy to get folded state
    std::ostringstream groovy;
    groovy << "def node = model.getNode('" << node_id_ << "');"
           << "node.isFolded()";
    
    auto result = client_->groovy(groovy.str());
    return result.find("true") != std::string::npos ||
           result.find("True") != std::string::npos ||
           result.find("1") != std::string::npos;
}

bool Node::setFolded(bool folded) {
    // Use Groovy to set folded state
    std::ostringstream groovy;
    groovy << "def node = model.getNode('" << node_id_ << "');"
           << (folded ? "node.fold()" : "node.unfold()");
    
    auto result = client_->groovy(groovy.str());
    return result.find("Error") == std::string::npos;
}

// -- actions -------------------------------------------------------------

bool Node::select() {
    return client_->focusNode(node_id_);
}

bool Node::center() {
    return client_->focusNode(node_id_);
}

void Node::refresh() {
    // Reload by fetching the text
    getText();
}

std::string Node::toString() const {
    std::ostringstream oss;
    oss << "Node(id=" << node_id_ << ")";
    try {
        oss << ", text=" << getText();
    } catch (const FreeplaneOperationError&) {
        oss << ", text=<unreadable>";
    }
    return oss.str();
}

}  // namespace grpc
}  // namespace freeplane
